File: jscomp/test/variant.ml
type pair = int * int

type v = | A | B of int | C of int * int | D of pair

let a = A

let b = B 34

let c = C (4,2)

let d = D (4,2)

let () = Js.log2 "a" a
let () = Js.log2 "b" b
let () = Js.log2 "c" c
let () = Js.log2 "d" d

let foo = function
| A -> 1
| B n -> n
| C (n,m) -> n+m
| D (n,m) -> n+m

File: jscomp/test/variant.js
'use strict';


var b = /* constructor */{
  tag: 0,
  name: "B",
  "0": 34
};

var c = /* constructor */{
  tag: 1,
  name: "C",
  "0": 4,
  "1": 2
};

var d = /* constructor */{
  tag: 2,
  name: "D",
  "0": /* tuple */[
    4,
    2
  ]
};

console.log("a", /* A */0);

console.log("b", b);

console.log("c", c);

console.log("d", d);

function foo(param) {
  if (typeof param === "number") {
    return 1;
  } else {
    switch (param.tag | 0) {
      case 0 :
          return param[0];
      case 1 :
          return param[0] + param[1] | 0;
      case 2 :
          var match = param[0];
          return match[0] + match[1] | 0;
      
    }
  }
}

var a = /* A */0;

exports.a = a;
exports.b = b;
exports.c = c;
exports.d = d;
exports.foo = foo;
/*  Not a pure module */
